src/Metadata/Cip25/Cip25String.purs

```diff
@@ -0,0 +1,145 @@
+-- | This module includes a string type that is used in CIP-25 standard.
+module Metadata.Cip25.Cip25String
+  ( Cip25String
+  , mkCip25String
+  , unCip25String
+  , fromMetadataString
+  , toMetadataString
+  , toCip25Strings
+  , fromCip25Strings
+  , fromDataString
+  , toDataString
+  ) where
+
+import Prelude
+
+import Aeson
+  ( class DecodeAeson
+  , class EncodeAeson
+  , JsonDecodeError(TypeMismatch)
+  , decodeAeson
+  )
+import Control.Alt ((<|>))
+import Data.Array ((:))
+import Data.Array as Array
+import Data.Either (hush, note)
+import Data.Foldable (fold)
+import Data.Maybe (Maybe(Nothing, Just), isJust)
+import Data.Newtype (unwrap, wrap)
+import Data.String.CodePoints as String
+import Data.TextDecoder (decodeUtf8)
+import Data.TextEncoder (encodeUtf8)
+import Data.Tuple.Nested (type (/\), (/\))
+import FromData (class FromData, fromData)
+import Metadata.FromMetadata (class FromMetadata, fromMetadata)
+import Metadata.ToMetadata (class ToMetadata, toMetadata)
+import ToData (class ToData, toData)
+import Types.ByteArray (ByteArray, byteLength)
+import Types.PlutusData (PlutusData)
+import Types.TransactionMetadata (TransactionMetadatum)
+
+-- | A string type that is used in CIP-25 standard.
+-- | String length in bytes (in UTF-8) is limited by 64, because PlutusData
+-- | bytes have this length limit.
+newtype Cip25String = Cip25String String
+
+derive newtype instance Eq Cip25String
+derive newtype instance Ord Cip25String
+derive newtype instance ToMetadata Cip25String
+derive newtype instance FromMetadata Cip25String
+derive newtype instance ToData Cip25String
+derive newtype instance FromData Cip25String
+derive newtype instance EncodeAeson Cip25String
+
+instance Show Cip25String where
+  show (Cip25String str) = "(unsafePartial (fromJust (mkCip25String "
+    <> show str
+    <> ")))"
+
+instance DecodeAeson Cip25String where
+  decodeAeson = decodeAeson >=> mkCip25String >>> note
+    (TypeMismatch "Cip25String")
+
+unCip25String :: Cip25String -> String
+unCip25String (Cip25String str) = str
+
+-- | A smart constructor for `Cip25String`
+mkCip25String :: String -> Maybe Cip25String
+mkCip25String str
+  | byteLength (wrap (encodeUtf8 str)) <= 64 = Just $ Cip25String str
+  | otherwise = Nothing
+
+takeCip25String :: String -> Maybe { init :: Cip25String, rest :: Maybe String }
```

Comment:
I appreciate the work put in to generalize with `forwardSearch` etc, however, I feel like this would be more simply implemented by repeatedly `Data.String.CodePoints.uncons` and applying `Data.String.CodePoints.fromCodePointArray` until failure to `mkCip25String`? Am I missing something?

Reply:
This is basically no different from `forwardSearch`, except that the step is set to 1.

I don't understand how CodePoints are related.

> CodePoint is an Int bounded between 0 and 0x10FFFF

Reply:
For the same reason you're using them here, we iterate using codepoints such that we don't break the encoding

Reply:
Also, I hadn't seen that step is being divided in the final case - this feels like a premature optimisation to me though

Reply:
Actually it seems like a mistake to me that you're using CodePoints in your implementation, right?

Reply:
Maybe it's not, to me this isn't straight forward to understand

Reply:
I think I understand it now, 16 point points might be exactly 64 bytes, but if it's not, it'll always be less, so we attempt take 24 codepoints (why this amount?) and try and fill up the rest of the remaining bytes with smaller and smaller amounts until it completes.

Reply:
(64-16) / 2 = 24

Reply:
> For the same reason you're using them here, we iterate using codepoints such that we don't break the encoding

Yes, but we don't need to access actual code points as numbers.

Reply:
> this feels like a premature optimisation to me though

It's making an algorithm that is quadratic in time to be `O(n * log(n))`.

Reply:
> Yes, but we don't need to access actual code points as numbers.
The number repr wouldn't be accessed, it's just so we never split a character incorrectly

Reply:
Then it does not matter if we use code point array slicing or string slicing

Reply:
Yeah, the suggestion wasn't from a correctness perspective, just an alternate implementation, sorry that wasn't clear. I think on average your implementation will have slightly better performance, it's unclear to me if that performance benefit is worth the extra code.

It's also worth bringing up, what are we supposed to do with combining characters/diacritics?

Reply:
I think that if the following property holds, we are safe:

```
forall ix str. 
  encodeUtf8 ((String.splitAt ix str).head) <> encodeUtf8 ((String.splitAt ix str).tail) == 
    encodeUtf8 str
```

Reply:
> It's making an algorithm that is quadratic in time to be `O(n * log(n))`

the `n` here is pretty small though, plus it's trivial to turn the quadratic algorithm into linear by using Data.String.CodeUnits.length to get how many bytes are in the codepoint (represented as a string).

I don't think what you've written is `O(n * log(n))` anyway, due to invoking `encodeUtf8` at each test?

If we really care about performance we can do better though, using `TextDecoder` we can take 64 bytes and decode them, count how many replacement characters at the end (the only place decoding can fail if the input was valid, and only up to 3), drop them, and use that information to aid in decoding the next. I'll create a mockup so you can decide for yourself

Reply:
I think this is about as fast as it gets in terms of big o, but also most operations are delegated to node/the browser, and it's in tail call form. Needs more testing, but what do you think?

https://github.com/Plutonomicon/cardano-transaction-lib/blob/df914937c0b36903f98150c1d150f5b34d20126f/src/Split64.purs#L26-L49

Reply:
Just to be clear: I'm happy with your current implementation, but maybe with a comment explaining the approach

Reply:
https://github.com/Plutonomicon/cardano-transaction-lib/blob/1c4afec40278ee21a62c20cee7f95f246ea07256/src/Split64.purs#L26-L27

@klntsky It's probably better we specify it using escape characters, it's a standard character used to replace invalid data in unicode streams. I seem to remember seeing a specification somewhere that TextDecoder will replace it, but there's at least some reference here https://infra.spec.whatwg.org/#javascript-string-convert.